### PR TITLE
Changes for Review

### PR DESCRIPTION
**Pull request by Rakeshrst**

### CODE REVIEW

File: src/main/java/com/rst/bootcamp/controller/BootCampController.java [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+		return instructorService.retrieveAllInstructors();
+	}
+
+	@PostMapping("/instructor")
+	public ResponseEntity<Instructor> saveInstructor(@RequestBody Instructor instructor) {
+		
+		instructor=instructorService.saveInstructor(instructor);
+		
+		URI location = ServletUriComponentsBuilder.fromCurrentRequest()
+                .path("/{id}")
```
Comment (asraniinfo) [inline src/main/java/com/rst/bootcamp/controller/BootCampController.java]: have used this "/{id}" in three different places in same class 
should create a global final variable an use at all places

File: src/main/java/com/rst/bootcamp/controller/BootCampController.java [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+                .buildAndExpand(instructor.getId())
+                .toUri();
+
+		
+		return ResponseEntity.created(location).build();
+	}
+	
+	@PostMapping("/instructor/{id}/subordinate")
+	public ResponseEntity<Void> addSubordinate(@PathVariable int id,@RequestBody Instructor subordinate) {
```
Comment (asraniinfo) [inline src/main/java/com/rst/bootcamp/controller/BootCampController.java]: here Instructor should be simple POJO or Data transfer objects instead of Entity exposed directly

Comment (asraniinfo) [reply src/main/java/com/rst/bootcamp/controller/BootCampController.java]: same for other method as well

File: src/main/java/com/rst/bootcamp/controller/BootCampController.java [hunk 3]
```diff
@@ -0,0 +1,80 @@
+package com.rst.bootcamp.controller;
+
+import java.net.URI;
+import java.util.List;
+
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.http.HttpStatus;
```
Comment (asraniinfo) [inline src/main/java/com/rst/bootcamp/controller/BootCampController.java]: Unused import
should clean code before PR is raised

File: src/main/java/com/rst/bootcamp/controller/BootCampController.java [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+		instructorService.addSubordinate(instructor, subordinate);
+		 instructor= instructorService.getInstructorDetails(id);
+		URI uri = ServletUriComponentsBuilder.fromCurrentRequest()
+				.path("/{id}").buildAndExpand(instructor.getId()).toUri();
+		
+		return ResponseEntity.created(uri).build();
+	}
+	
+	@PostMapping("/instructor/{id}/subordinate/{subordinate_id}")
+	public ResponseEntity<Void> addSubordinate(@PathVariable int id,@PathVariable int subordinate_id) {
```
Comment (asraniinfo) [inline src/main/java/com/rst/bootcamp/controller/BootCampController.java]: improper naming   
-subordinate_id
+subordinateId

File: src/test/java/com/rst/bootcamp/controller/BootCampControllertest.java [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+import org.springframework.boot.test.autoconfigure.web.servlet.WebMvcTest;
+import org.springframework.boot.test.mock.mockito.MockBean;
+import org.springframework.http.MediaType;
+import org.springframework.test.web.servlet.MockMvc;
+
+import com.rst.bootcamp.model.Instructor;
+import com.rst.bootcamp.service.InstructorService;
+
+@WebMvcTest(BootCampController.class)
+class BootCampControllertest {
```
Comment (asraniinfo) [inline src/test/java/com/rst/bootcamp/controller/BootCampControllertest.java]: improper naming
-BootCampControllertest
+BootCampControllerTest

File: src/test/java/com/rst/bootcamp/BootcampApplicationTests.java [hunk 6]
```diff
@@ -5,9 +5,7 @@
 
 @SpringBootTest
 class BootcampApplicationTests {
-
 	@Test
 	void contextLoads() {
```
Comment (asraniinfo) [inline src/test/java/com/rst/bootcamp/BootcampApplicationTests.java]: No Tests

add tests if can't remove this empty code.

File: src/main/java/com/rst/bootcamp/service/InstructorService.java [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+	public Instructor saveInstructor(Instructor instructor){
+		
+		instructor= instructorRepo.save(instructor);
+		
+		return instructor;
+	}
+	
+	public Instructor getInstructorDetails(int id){
+		
+		Instructor instructor= instructorRepo.findById(id).get();
```
Comment (asraniinfo) [inline src/main/java/com/rst/bootcamp/service/InstructorService.java]: this is possible null pointer
after getting this instructorRepo.findById(id)  check for null before get

File: src/main/java/com/rst/bootcamp/service/InstructorService.java [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+		
+		return instructor;
```
Comment (asraniinfo) [inline src/main/java/com/rst/bootcamp/service/InstructorService.java]: unnecessary assignment to variable 
could directly return instructorRepo.findById(id).get();

File: src/test/java/com/rst/bootcamp/InstructorsIntegrationTests.java [hunk 9]
```diff
@@ -0,0 +1,142 @@
...
+					.json("{id: 1,instructorName: I1,adminName: A1,subordinate:{id: 2,instructorName: I2,adminName: A1}}"))
+					.andExpect(status().isOk()).andReturn();
+
+		} catch (Throwable e) {
+
+		}
+	}
+	
+	@Test
+	void contextLoads() {
```
Comment (asraniinfo) [inline src/test/java/com/rst/bootcamp/InstructorsIntegrationTests.java]: empty test.

**Comment by asraniinfo** [comment]
Please check for warning and run any of static analysis before submitting code(sonar/coverity)

File: src/test/java/com/rst/bootcamp/InstructorsIntegrationTests.java [hunk 10]
```diff
@@ -0,0 +1,142 @@
...
+public class InstructorsIntegrationTests {
+
+	@Autowired
+	private MockMvc mockMvc;
+
+	@Autowired
+	InstructorRepository instructorRepo;
+
+	public static final MediaType APPLICATION_JSON_UTF8 = new MediaType(MediaType.APPLICATION_JSON.getType(),
+			MediaType.APPLICATION_JSON.getSubtype(), Charset.forName("utf8"));
```
Comment (asraniinfo) [inline src/test/java/com/rst/bootcamp/InstructorsIntegrationTests.java]: use StandardCharsets.UTF_8 instead

File: src/main/java/com/rst/bootcamp/model/Instructor.java [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+
+	@Id
+	@GeneratedValue
+	private int id;
+
+	private String instructorName;
+
+	private String adminName;
+
+	@OneToOne(cascade = CascadeType.ALL, fetch = FetchType.EAGER, optional = true)
```
Comment (asraniinfo) [inline src/main/java/com/rst/bootcamp/model/Instructor.java]: classic tree structure
can have multiple subordinates, change to  OneToMany

Comment (asraniinfo) [reply src/main/java/com/rst/bootcamp/model/Instructor.java]: optional = true is redundant 
not required if true

File: src/test/java/com/rst/bootcamp/InstructorsIntegrationTests.java [hunk 12]
```diff
@@ -0,0 +1,142 @@
...
+	public static String asJsonString(final Object obj) {
+		try {
+			return new ObjectMapper().writeValueAsString(obj);
+		} catch (Exception e) {
+			throw new RuntimeException(e);
+		}
+	}
+
+	@Test
+	public void test_addSubordinate() throws Exception {
```
Comment (asraniinfo) [inline src/test/java/com/rst/bootcamp/InstructorsIntegrationTests.java]: declaration of throws Exception is not required,
Exception is never thrown

File: src/test/java/com/rst/bootcamp/InstructorsIntegrationTests.java [hunk 13]
```diff
@@ -0,0 +1,142 @@
...
+							.contentType(APPLICATION_JSON_UTF8))
+					.andExpect(status().isCreated());
+
+		} catch (Throwable e) {
+
+		}
+	}
+
+	@Test
+	public void test_getAllInstructors() throws Exception {
```
Comment (asraniinfo) [inline src/test/java/com/rst/bootcamp/InstructorsIntegrationTests.java]: same as line 57

File: src/test/java/com/rst/bootcamp/InstructorsIntegrationTests.java [hunk 14]
```diff
@@ -0,0 +1,142 @@
...
+
+		}
+	}
+	
+	@Test
+	void contextLoads() {
+	}
+	
+	/**
+	@Test
```
Comment (asraniinfo) [inline src/test/java/com/rst/bootcamp/InstructorsIntegrationTests.java]: should never leave commented code.

remove

File: src/test/java/com/rst/bootcamp/InstructorsIntegrationTests.java [hunk 15]
```diff
@@ -0,0 +1,142 @@
...
+					.andExpect(status().isCreated());
+
+			List<Instructor> instructors = instructorRepo.findAll();
+
+			mockMvc.perform(
+					post("/instructor/" + instructors.get(0).getId() + "/subordinate/" + instructors.get(1).getId())
+							.contentType(APPLICATION_JSON_UTF8))
+					.andExpect(status().isCreated());
+
+		} catch (Throwable e) {
```
Comment (asraniinfo) [inline src/test/java/com/rst/bootcamp/InstructorsIntegrationTests.java:66]: empty catch block
do something.

**Review by asraniinfo — CHANGES_REQUESTED** [review]

File: src/main/java/com/rst/bootcamp/controller/BootCampController.java (hunk 2 above)
Comment (Rakeshrst) [reply src/main/java/com/rst/bootcamp/controller/BootCampController.java]: done

File: src/main/java/com/rst/bootcamp/controller/BootCampController.java (hunk 3 above)
Comment (Rakeshrst) [reply src/main/java/com/rst/bootcamp/controller/BootCampController.java]: Done

File: src/main/java/com/rst/bootcamp/controller/BootCampController.java (hunk 4 above)
Comment (Rakeshrst) [reply src/main/java/com/rst/bootcamp/controller/BootCampController.java]: changed the method to accept this by name and removed additon using id

File: src/test/java/com/rst/bootcamp/controller/BootCampControllertest.java (hunk 5 above)
Comment (Rakeshrst) [reply src/test/java/com/rst/bootcamp/controller/BootCampControllertest.java]: Done

File: src/main/java/com/rst/bootcamp/model/Instructor.java (hunk 11 above)
Comment (Rakeshrst) [reply src/main/java/com/rst/bootcamp/model/Instructor.java]: Did the tree structure.

File: src/test/java/com/rst/bootcamp/InstructorsIntegrationTests.java (hunk 12 above)
Comment (Rakeshrst) [reply src/test/java/com/rst/bootcamp/InstructorsIntegrationTests.java]: Understood

File: src/test/java/com/rst/bootcamp/InstructorsIntegrationTests.java (hunk 14 above)
Comment (Rakeshrst) [reply src/test/java/com/rst/bootcamp/InstructorsIntegrationTests.java]: Understood.

File: src/test/java/com/rst/bootcamp/InstructorsIntegrationTests.java (hunk 15 above)
Comment (Rakeshrst) [reply src/test/java/com/rst/bootcamp/InstructorsIntegrationTests.java:66]: Understood

File: README.md [hunk 16]
```diff
@@ -1 +1,2 @@
 # bootcamp
```
Comment (ankit-agrawal1) [inline README.md:1]: add execution steps 
aka. ussage/installation steps

File: src/main/java/com/rst/bootcamp/model/Instructor.java [hunk 17]
```diff
@@ -0,0 +1,124 @@
...
+	@Column(unique=true)
+	private String instructorName;
+
+
+	  @JsonIgnore
+	  @ManyToOne(cascade = { CascadeType.PERSIST})
+	  private Instructor parent;
+	  
+	  @Transient
+	  private String admin;
```
Comment (ankit-agrawal1) [inline src/main/java/com/rst/bootcamp/model/Instructor.java:36]: this should be of Instructor type.

**Review by ankit-agrawal1 — CHANGES_REQUESTED** [review]

File: src/main/java/com/rst/bootcamp/model/Instructor.java (hunk 17 above)
Comment (Rakeshrst) [reply src/main/java/com/rst/bootcamp/model/Instructor.java:36]: I have the parent as the type Instructor - and admin is just a field to show.  
If I keep Admin as the type Instructor. The whole Response structure keeps on going on a never ending loop.
Still looking how can I make it possible.

**Comment by ankit-agrawal1** [comment]
While running application getting this exception


org.springframework.boot.devtools.restart.SilentExitExceptionHandler$SilentExitException
org.springframework.boot.devtools.restart.SilentExitExceptionHandler$SilentExitException
	at org.springframework.boot.devtools.restart.SilentExitExceptionHandler.exitCurrentThread(SilentExitExceptionHandler.java:92)
	at org.springframework.boot.devtools.restart.Restarter.immediateRestart(Restarter.java:180)
	at org.springframework.boot.devtools.restart.Restarter.initialize(Restarter.java:164)
	at org.springframework.boot.devtools.restart.Restarter.initialize(Restarter.java:554)
	at org.springframework.boot.devtools.restart.RestartApplicationListener.onApplicationStartingEvent(RestartApplicationListener.java:90)
	at org.springframework.boot.devtools.restart.RestartApplicationListener.onApplicationEvent(RestartApplicationListener.java:50)
	at org.springframework.context.event.SimpleApplicationEventMulticaster.doInvokeListener(SimpleApplicationEventMulticaster.java:176)
	at org.springframework.context.event.SimpleApplicationEventMulticaster.invokeListener(SimpleApplicationEventMulticaster.java:169)
	at org.springframework.context.event.SimpleApplicationEventMulticaster.multicastEvent(SimpleApplicationEventMulticaster.java:143)
	at org.springframework.context.event.SimpleApplicationEventMulticaster.multicastEvent(SimpleApplicationEventMulticaster.java:131)
	at org.springframework.boot.context.event.EventPublishingRunListener.starting(EventPublishingRunListener.java:76)
	at org.springframework.boot.SpringApplicationRunListeners.lambda$starting$0(SpringApplicationRunListeners.java:53)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1541)
	at org.springframework.boot.SpringApplicationRunListeners.doWithListeners(SpringApplicationRunListeners.java:117)
	at org.springframework.boot.SpringApplicationRunListeners.starting(SpringApplicationRunListeners.java:53)
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:317)
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:1313)
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:1302)
	at com.rst.bootcamp.BootcampApplication.main(BootcampApplication.java:10)

**Comment by Rakeshrst** [comment]
> While running application getting this exception
> 
> org.springframework.boot.devtools.restart.SilentExitExceptionHandler$SilentExitException
> org.springframework.boot.devtools.restart.SilentExitExceptionHandler$SilentExitException
> at org.springframework.boot.devtools.restart.SilentExitExceptionHandler.exitCurrentThread(SilentExitExceptionHandler.java:92)
> at org.springframework.boot.devtools.restart.Restarter.immediateRestart(Restarter.java:180)
> at org.springframework.boot.devtools.restart.Restarter.initialize(Restarter.java:164)
> at org.springframework.boot.devtools.restart.Restarter.initialize(Restarter.java:554)
> at org.springframework.boot.devtools.restart.RestartApplicationListener.onApplicationStartingEvent(RestartApplicationListener.java:90)
> at org.springframework.boot.devtools.restart.RestartApplicationListener.onApplicationEvent(RestartApplicationListener.java:50)
> at org.springframework.context.event.SimpleApplicationEventMulticaster.doInvokeListener(SimpleApplicationEventMulticaster.java:176)
> at org.springframework.context.event.SimpleApplicationEventMulticaster.invokeListener(SimpleApplicationEventMulticaster.java:169)
> at org.springframework.context.event.SimpleApplicationEventMulticaster.multicastEvent(SimpleApplicationEventMulticaster.java:143)
> at org.springframework.context.event.SimpleApplicationEventMulticaster.multicastEvent(SimpleApplicationEventMulticaster.java:131)
> at org.springframework.boot.context.event.EventPublishingRunListener.starting(EventPublishingRunListener.java:76)
> at org.springframework.boot.SpringApplicationRunListeners.lambda$starting$0(SpringApplicationRunListeners.java:53)
> at java.base/java.util.ArrayList.forEach(ArrayList.java:1541)
> at org.springframework.boot.SpringApplicationRunListeners.doWithListeners(SpringApplicationRunListeners.java:117)
> at org.springframework.boot.SpringApplicationRunListeners.starting(SpringApplicationRunListeners.java:53)
> at org.springframework.boot.SpringApplication.run(SpringApplication.java:317)
> at org.springframework.boot.SpringApplication.run(SpringApplication.java:1313)
> at org.springframework.boot.SpringApplication.run(SpringApplication.java:1302)
> at com.rst.bootcamp.BootcampApplication.main(BootcampApplication.java:10)

Can you try this ? 
mvn spring-boot:run -Dspring.devtools.restart.enabled=false

**Comment by Rakeshrst** [comment]
> > While running application getting this exception
> > org.springframework.boot.devtools.restart.SilentExitExceptionHandler$SilentExitException
> > org.springframework.boot.devtools.restart.SilentExitExceptionHandler$SilentExitException
> > at org.springframework.boot.devtools.restart.SilentExitExceptionHandler.exitCurrentThread(SilentExitExceptionHandler.java:92)
> > at org.springframework.boot.devtools.restart.Restarter.immediateRestart(Restarter.java:180)
> > at org.springframework.boot.devtools.restart.Restarter.initialize(Restarter.java:164)
> > at org.springframework.boot.devtools.restart.Restarter.initialize(Restarter.java:554)
> > at org.springframework.boot.devtools.restart.RestartApplicationListener.onApplicationStartingEvent(RestartApplicationListener.java:90)
> > at org.springframework.boot.devtools.restart.RestartApplicationListener.onApplicationEvent(RestartApplicationListener.java:50)
> > at org.springframework.context.event.SimpleApplicationEventMulticaster.doInvokeListener(SimpleApplicationEventMulticaster.java:176)
> > at org.springframework.context.event.SimpleApplicationEventMulticaster.invokeListener(SimpleApplicationEventMulticaster.java:169)
> > at org.springframework.context.event.SimpleApplicationEventMulticaster.multicastEvent(SimpleApplicationEventMulticaster.java:143)
> > at org.springframework.context.event.SimpleApplicationEventMulticaster.multicastEvent(SimpleApplicationEventMulticaster.java:131)
> > at org.springframework.boot.context.event.EventPublishingRunListener.starting(EventPublishingRunListener.java:76)
> > at org.springframework.boot.SpringApplicationRunListeners.lambda$starting$0(SpringApplicationRunListeners.java:53)
> > at java.base/java.util.ArrayList.forEach(ArrayList.java:1541)
> > at org.springframework.boot.SpringApplicationRunListeners.doWithListeners(SpringApplicationRunListeners.java:117)
> > at org.springframework.boot.SpringApplicationRunListeners.starting(SpringApplicationRunListeners.java:53)
> > at org.springframework.boot.SpringApplication.run(SpringApplication.java:317)
> > at org.springframework.boot.SpringApplication.run(SpringApplication.java:1313)
> > at org.springframework.boot.SpringApplication.run(SpringApplication.java:1302)
> > at com.rst.bootcamp.BootcampApplication.main(BootcampApplication.java:10)
> 
> Can you try this ?
> mvn spring-boot:run -Dspring.devtools.restart.enabled=false

Are you using IntelliJ or Eclipse ?
